File: go/p14.go
package main

func collatzSequenceLength(num int) int {
	countTerms := 1

	for num > 1 {
		if num%2 == 0 {
			num = num / 2
		} else {
			num = 3*num + 1
		}
		countTerms++
	}

	return countTerms
}

func p14() int {
	var (
		currentLength int
		longestLength int
		startNum int
	)

	for i := 0; i < 1000000; i++ {
		currentLength = collatzSequenceLength(i)

		if currentLength > longestLength {
			longestLength = currentLength
			startNum = i
		}

	}
	return startNum
}

File: cpp/p14.cpp
#include <iostream>

int collatzSequenceLength(int num)
{
    int length = 1;
    while (num != 1)
    {
        if (num % 2 == 0)
        {
            num = num / 2;
        }
        else
        {
            num = 3 * num + 1;
        }
        length++;
    }

    return length;
}

int main(int argc, char const *argv[])
{
    int currentLength, longestLength = 0, startNum = 0;
    for (size_t i = 0; i < 100000; i++)
    {
        currentLength = collatzSequenceLength(i);
        if (currentLength >= longestLength)
        {
            currentLength = longestLength;
            startNum = i;
        }
    }

    std::cout << startNum << std::endl;

    return 0;
}
